### DISCUSSION TITLE
Release build in watch mode

### DISCUSSION THREAD

**Discussion by 7tupel** [thheller/shadow-cljs#1131] [General]
Hi, 

is there any way to create a release build while watching for code changes and automatically rebuild the release?

Now, this might seem contra intuitive on first sight. But I have a usecase with an external dev environment that requires a full JS build to be used during development. (The environment connects to the local development machine through a proprietary tunnel and pushes the changed code from local to the remote.)  On the remote environment `unsafe-eval` is not allowed, but the normal development build does contain text that gets evaluated this way:
```
EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'
```

It is not possible to change anything in this general setup unfortunately.

It would of cause be possible to run some script in the background and watch for changes in the `src` directory and run a `shadow-cljs release app` but it would be a lot nicer if this would be possible directly from shadow.

Disclaimer: Yes, I know this is not nice. If I had a choice I would do it differently but the platform/environment the app will be deployed in is (still) not very great in terms of development experience.

**Comment by thheller** [comment]
No, there is no built-in way to have an automatically triggered `release` build.

You could try setting `:devtools {:loader-mode :script}` in the build config? This removes some eval, maybe that works?

Did you consider using a local server and proxying requests to the remote env? https://shadow-cljs.github.io/docs/UsersGuide.html#dev-http-proxy

I assume this is using `:target :browser`? You could also try `:target :esm`, as it does not `eval` at all.

**Comment by 7tupel** [comment]
Unfortunately the variant with `:devtools {:loader-mode :script}` does not fix the issue. I guess the environment does not allow any `eval` to be present in the JS code.

Not sure if esm will work. The app is basically a React app that gets mounted inside an iframe. I have to provide an `index.html` file that defines a div element inside of body that is used to mount the React app. The whole index.html file is what gets loaded inside the iframe.

I don't know if it even possible to bundle a React app as ESM and then load and initialise the app from an iframe in Atlassian Cloud.

**Reply by 7tupel** [reply]
I'm btw talking about the Atlassian Cloud platform. It is very limited and not everything is documented as well as I would like as a developer... I'm trying to create a PoC to build Atlassian Cloud Apps using Clojure(-Script) instead of JS.

**Reply by thheller** [reply]
`react` works just fine in an `:esm` build. Just need to load the script with `type="module"`. If its just a regular iframe that should be totally fine.

Unfortunately I don't know a single thing about Atlassian Cloud Apps, so can't make any other suggestions.

**Reply by 7tupel** [reply]
It seems I got it working or at least some of it. The trick was to import a "normal" js file from the iframe's html that in turn imports and re-exports the react app. 

The atlassian toolchain is closed source and does who knows what to compile and bundle code. This has given us a lot of headaches in the past and will continue to do so... :/ 

Thanks a lot for your support :)
